chain/chain/src/chain.rs

```diff
@@ -1334,17 +1341,17 @@ impl Chain {
     /// these blocks that are ready.
     /// `block_processing_artifacts`: Callers can pass an empty object or an existing BlockProcessingArtifact.
     ///              This function will add the effect from processing this block to there.
-    /// `apply_chunks_done_callback`: This callback will be called after apply_chunks are finished
+    /// `apply_chunks_done_sender`: An ApplyChunksDoneMessage message will be send via this sender after apply_chunks is finished
     ///              (so it also happens asynchronously in the rayon thread pool). Callers can
-    ///              use this callback as a way to receive notifications when apply chunks are done
+    ///              use this sender as a way to receive notifications when apply chunks are done
     ///              so it can call postprocess_ready_blocks.
     pub fn start_process_block_async(
         &mut self,
         me: &Option<AccountId>,
         block: MaybeValidated<Block>,
         provenance: Provenance,
         block_processing_artifacts: &mut BlockProcessingArtifact,
-        apply_chunks_done_callback: DoneApplyChunkCallback,
+        apply_chunks_done_sender: Option<near_async::messaging::Sender<ApplyChunksDoneMessage>>,
```

Comment:
Why does sender need to be passed in as an argument?

Reply:
The callback basically sends a message to ClientActor to notify it the end of the apply-chunks (to it does processing of unfinished blocks). As far as I can see, the chain code (which eventually sends this message) does not have access to the client, so we need to pass through the sender to enable this mechanism. (well, I am no quite happy with this change as there is a lot of propagation of this parameter through a long call stack.)

did I understand your question right?

Reply:
We can clone the channel and make it a field of `Chain` if that helps.

Reply:
I considered keeping the sender as part of Chain, but it will not be less nontrivial, since the sender belongs to ClientAction and it should be passed to Chain during construction (from ClientAction to ClientActor, to Client, and to Chain) Chain construction is currently is sort of independent of Client/ClientActor construction (Client constructs Chain in production code but tests do the construction independently and also does not rely on existence of ClientAction or ClientActor). Thus, I prefer for now exposing the sender explicitly instead of the callback, and consider refactoring it later.